### PR TITLE
[GSoC2024] resolved issue #7443 (navigate over unresolved issues)

### PR DESCRIPTION
**Pull request by thekavikumar**
Resolved issue #7443 

It skips the frames where all issues have been resolved!

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have created a changelog fragment 
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

**Comment by thekavikumar** [comment]
@klakhov please review this pr,  I've made the required changes

**Comment by thekavikumar** [comment]
@nmanovic @klakhov a reminder, i request you to review my pr

**Comment by nmanovic** [comment]
@thekavikumar , could you please take a look at the failed test?

**Comment by thekavikumar** [comment]
@nmanovic let me check

File: cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/issues-list.tsx
Comment (klakhov) [inline cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/issues-list.tsx:43]: The filter should only be applied if the hide issues setting enabled

Comment (thekavikumar) [reply cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/issues-list.tsx:43]: alright

**Comment by thekavikumar** [comment]
@klakhov please check it now

File: cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/issues-list.tsx
Comment (klakhov) [inline cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/issues-list.tsx:45]: We do not need comments here, the code is pretty self-explanatory

File: changelog.d/20240321_205232_kavikumarceo_resolved_issue_7443.md
Comment (klakhov) [inline changelog.d/20240321_205232_kavikumarceo_resolved_issue_7443.md:3]: I belive its not an `added a feature` but fix a bug, The filter should've been skipping issues that are hidden

Comment (klakhov) [inline changelog.d/20240321_205232_kavikumarceo_resolved_issue_7443.md:3]:
```suggestion
### Fixed

- `Find next frame with issues` ignored `hide resolved issues` setting  
```

**Review by klakhov — COMMENTED** [review]
There is a problem with tests, please look on failed ones.

Also, it seems the fix is not working(filter is always disabled):
issues

File: tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js
Comment (klakhov) [inline tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js:287]: Why do we need this change? If we force the clicks it prevents cypress from checking if the button is [actionable](https://docs.cypress.io/guides/core-concepts/interacting-with-elements#Forcing), which is not what we really want here.

**Review by klakhov — COMMENTED** [review]
Hi @thekavikumar, 
The fix seems to be working now. Do you mind to add a cypress test for it? We can add checks for this setting in review_pipeline_feature issue navigation part.

File: tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js
Comment (thekavikumar) [reply tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js:287]: @klakhov sure I'll add a cypress test

Comment (thekavikumar) [reply tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js:287]: `it('Calculates frames correctly based on issues and conflicts', () => {
    // Mock state to simulate Redux store
    const state = {
        annotation: {
            player: {
                frame: {
                    number: 3 // Current frame number
                }
            },
            workspace: Workspace.REVIEW,
            canvas: {
                ready: true,
                activeControl: ActiveControl.CURSOR
            }
        },
        review: {
            issues: [
                { id: 1, frame: 1, resolved: false },
                { id: 2, frame: 3, resolved: true },
                { id: 3, frame: 5, resolved: false }
            ],
            conflicts: [
                { id: 4, frame: 2 },
                { id: 5, frame: 4 }
            ]
        },
        settings: {
            shapes: {
                showGroundTruth: true
            }
        },
        // Other required state properties...
    };

    cy.wrap(state).then((state) => {
        cy.window().its('store').invoke('replaceReducer', () => state);

        // Ensure frames are calculated correctly
        cy.get('@dynamicLeftProps').then((dynamicLeftProps) => {
            const onClick = dynamicLeftProps.onClick;
            const nearestLeftFrame = onClick.toString().match(/\d+/g)[0];
            expect(nearestLeftFrame).to.equal('2'); // Nearest left frame should be 2
        });

        cy.get('@dynamicRightProps').then((dynamicRightProps) => {
            const onClick = dynamicRightProps.onClick;
            const nearestRightFrame = onClick.toString().match(/\d+/g)[0];
            expect(nearestRightFrame).to.equal('4'); // Nearest right frame should be 4
        });
    });
});
`

Is this test okay @klakhov ?

Comment (klakhov) [reply tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js:287]: @thekavikumar,
No, its not what we expect here. Consider looking on [review pipeline test](https://github.com/opencv/cvat/blob/develop/tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js). We use cypress to test UI from user perspective. Which means, we dont emulate state nor check components props. You can extend the review test with resolving some issues,setting hide resolved issues and check that navigation is correct as the normal user would do. We already test basic issue filter navigation on [280 line and below](https://github.com/opencv/cvat/blob/develop/tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js#L280)

**Comment by thekavikumar** [comment]
@klakhov by any chance can we run tests locally?

**Comment by thekavikumar** [comment]
@nmanovic @klakhov If there is no failing tests, I guess its good to merge now. What do you think sir?

**Comment by klakhov** [comment]
@thekavikumar 

> by any chance can we run tests locally?

You can run tests locally. [Refer to the tests documentation](https://opencv.github.io/cvat/docs/contributing/running-tests/). You will need first section with E2E tests.

> If there are no failing tests, I guess its good to merge now. What do you think sir?

Please, check my comment above where I've asked you to write a test, its good practice to add one with bugfix. I've [shared some ideas](https://github.com/opencv/cvat/pull/7653#discussion_r1540653809) about how it should be implemented.

**Comment by thekavikumar** [comment]
> @thekavikumar
> 
> > by any chance can we run tests locally?
> 
> You can run tests locally. [Refer to the tests documentation](https://opencv.github.io/cvat/docs/contributing/running-tests/). You will need first section with E2E tests.
> 
> > If there are no failing tests, I guess its good to merge now. What do you think sir?
> 
> Please, check my comment above where I've asked you to write a test, its good practice to add one with bugfix. I've [shared some ideas](https://github.com/opencv/cvat/pull/7653#discussion_r1540653809) about how it should be implemented.

Yeah will commit with a test written for soon!

**Comment by thekavikumar** [comment]
image

I'm getting this error, may I know why @klakhov

**Comment by thekavikumar** [comment]
@klakhov also I'm not getting what to test, because the change I did is already being tested [here ](https://github.com/opencv/cvat/blob/develop/tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js#L280) right?

**Comment by thekavikumar** [comment]
@klakhov can please make me get a clear idea of what to test with the bugfix?

**Comment by klakhov** [comment]
@thekavikumar 
Probably you get 400 error because the users are already created in the database from previous runs. Try to run `remove_users_tasks_projects_organizations.js` to delete them

The test that Ive linked test only basic navigation for frames with issues. It doesnt test filter setting. You need to resolve some issues, enable the filter and check that frames with resolved issues are not shown using the filter

**Comment by thekavikumar** [comment]
@klakhov okay so in the tests I have to write code to create multiple frames with issues, then resolve the issues in one frame, then try to navigate and check the frame number weather the frame has been skipped or not. Am I right?

**Comment by thekavikumar** [comment]
@klakhov should I create test inside this itself or should I create a separate describe?
image

**Comment by klakhov** [comment]
@thekavikumar 
I think you can extend the current review test after the part where we test usual issue filter. You can do such steps:
1. Resolve the issue on the frame in the middle(frame 1) (we have issues on 0, 1, 2 frames)
2. Enable the filter
3. Check that the frame 1 is skipped using the filter navigation
4. Check that full test works because we have resolving issue logic after the filter testing and fix that if needed

**Comment by thekavikumar** [comment]
@klakhov  how can i get all issues like this
image

for filtering

let issuesResolvedHidden = true;

// Filter frames based on resolved issues and store them in the 'frames' array
let frames = issues
    .filter((issue) => !issuesResolvedHidden || !issue.resolved)
    .map((issue) => issue.frame)
    .sort((a, b) => a - b);

// Resolve the issue on frame 1
cy.goCheckFrameNumber(1);
cy.collectIssueLabel().then((issueLabelList) => {
    for (const label of issueLabelList) {
        cy.resolveIssue(label, 'Resolved issue on frame 1');
    }
});

issuesResolvedHidden = false; 
cy.goCheckFrameNumber(0);
cy.get('.cvat-issues-sidebar-previous-frame')
                .should('have.attr', 'style')
                .and('contain', 'opacity: 0.5;'); // the element is not active
            cy.get('.cvat-issues-sidebar-next-frame').should('be.visible').click();
            cy.checkFrameNum(2);

**Comment by thekavikumar** [comment]
@klakhov does this code look good?

**Comment by thekavikumar** [comment]
@klakhov also when i run remove_user that test file i get 
image

**Comment by thekavikumar** [comment]
@klakhov can you please review it

**Comment by thekavikumar** [comment]
@klakhov i mean review the tests, i have committed it :)

**Comment by thekavikumar** [comment]
@klakhov sir one more thing, I'm not able to run full tests as because due to some reason when I try to remove users by running the [remove_users_tasks_projects_organizations.js](https://github.com/thekavikumar/cvat/blob/develop/tests/cypress/e2e/remove_users_tasks_projects_organizations.js)

could you please help me with that?

File: tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js
Comment (klakhov) [inline tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js:294]: We already have `countIssuesByFrame` for such purpose, why we need to add one more?

File: tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js
Comment (klakhov) [inline tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js:304]: It seems the for loop is viable only for frame == 1, so we dont need to iterate all over the frames.
We can just:
```
cy.goCheckFrameNumber(1);
cy.collectIssueLabel().then((issueLabelList) => {
                        for (let label = 0; label < issueLabelList.length; label++) {
                            cy.resolveIssue(issueLabelList[label], 'Resolved issue');
                        }
                    });
```

**Comment by klakhov** [comment]
As for invalid token error, please provide mode details. On what step the error happens? Do you run test from admin account?

**Comment by thekavikumar** [comment]
@klakhov yes I'm using the admin username and password only. i got this error while running 
image
image

File: tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js
Comment (thekavikumar) [reply tests/cypress/e2e/actions_users/registration_involved/case_28_review_pipeline_feature.js:304]: got it

**Comment by klakhov** [comment]
@thekavikumar 
Check the `id` of your admin user that you are using to run tests. Maybe its not `1` as expected in the delete test, so the user gets deleted and token is wrong. Change it inside the test if needed. 
Also it seems the `review` test is failing on the CI.

**Comment by thekavikumar** [comment]
@klakhov okay let me check

**Comment by thekavikumar** [comment]
@klakhov please give approval for the workflow

**Comment by thekavikumar** [comment]
@nmanovic @klakhov is there any possibility to run the entire tests in our local system as it runs in GitHub workflow because each time I have to push the changes then the admin have to approve for the workflow to run.

**Comment by thekavikumar** [comment]
@nmanovic @klakhov Sorry for asking you again for workflow approval. I've no other options. Please help me with this testing issues

**Comment by klakhov** [comment]
@thekavikumar 
I've approved your runs, but I see one problem in the test. After you resolved an issue on frame 1, the code on line 329 is trying to do the same and fails on frame 1 as its already resolved, please fix this.
Also, please try to run the test locally before submitting. Do you still have a problem with `remove_users_tasks_projects_organizations.js`?

**Comment by thekavikumar** [comment]
@klakhov 'After you resolved an issue on frame 1, the code on line 329 is trying to do the same and fails on frame 1 as its already resolved, please fix this.' for this I am asking, shall we reopen or shall I create new frames for this?

**Comment by thekavikumar** [comment]
@klakhov 
image
image

image
image

**Comment by klakhov** [comment]
You dont need to reopen or create new frames. The code on 329 tests issue resolving process, so if the issue is already resolved, we dont need to do anything, we need to skip resolving it.

**Comment by thekavikumar** [comment]
@klakhov okay let me try it

**Comment by thekavikumar** [comment]
@klakhov review the recent commit, will this work?

**Comment by klakhov** [comment]
You need to update server dependencies after merging develop

**Comment by thekavikumar** [comment]
@klakhov how do i update?

**Comment by thekavikumar** [comment]
@klakhov I'm not sure why this occured, any idea? Also can you approve the workflow once again?
image

**Comment by thekavikumar** [comment]
@klakhov the filter test was running good, but the above issue in image was in actions_projects_models, not sure why.

**Review by klakhov — APPROVED** [review]

**Comment by thekavikumar** [comment]
> You need to update server dependencies after merging develop

@klakhov @nmanovic may i know how to update the server deps?

**Comment by klakhov** [comment]
> @klakhov how do i update? using yarn?

To update server dependencies([link to docs](https://docs.cvat.ai/docs/contributing/development-environment/#setup-the-dependencies)): 
```
pip install -U pip wheel setuptools
pip install -r cvat/requirements/development.txt
```

**Comment by thekavikumar** [comment]
> > @klakhov how do i update? using yarn?
> 
> To update server dependencies([link to docs](https://docs.cvat.ai/docs/contributing/development-environment/#setup-the-dependencies)):
> 
> ```
> pip install -U pip wheel setuptools
> pip install -r cvat/requirements/development.txt
> ```

Alright that works, thanks

**Pull request merged**
